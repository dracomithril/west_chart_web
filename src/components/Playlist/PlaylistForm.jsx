import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Button, Fab } from '@material-ui/core';
import TextField from '@material-ui/core/TextField';
import Checkbox from '@material-ui/core/Checkbox';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSave, faLightbulb } from '@fortawesome/free-solid-svg-icons';
import { weekInfo } from '../../utils/utils';
import './playlist.css';

export class PlaylistForm extends Component {
  constructor(props) {
    super(props);
    this.state = {
      playlistName: '',
      isPrivate: false,
    };
  }

  onGenPlaylistName = () => {
    const { monday, friday } = weekInfo();

    const mondayStr = monday.format('MMM_D').toUpperCase();
    const fridayStr = friday.format('MMM_D').toUpperCase();

    const playlistName = `Chart_${mondayStr}-${fridayStr}`;
    this.setState({ playlistName });
  };

  render() {
    const {
      onCreatePlaylistClick: onCreatePlaylist, onStartClick, hasElements, isUserLogged, spotifyUser,
    } = this.props;
    const { playlistName, isPrivate } = this.state;
    const disable_create = !(playlistName.length > 5 && hasElements && isUserLogged);
    const validatePlaylistName = (str = '') => str.length > 7;
    const isNameValid = validatePlaylistName(playlistName);
    return (
      <div className="playlist-form">
        <Button variant="contained" color="primary" onClick={onStartClick} id="start_sp_button">
          start
        </Button>
        <TextField
          error={!isNameValid}
          value={playlistName}
          style={{ padding: '5px' }}
          placeholder="playlist name"
          onChange={({ target }) => {
            this.setState({ playlistName: target.value });
          }}
        />
        <div style={{ alignSelf: 'center' }}>
          <Fab color="secondary" onClick={this.onGenPlaylistName} id="genName_sp_button">
            <FontAwesomeIcon icon={faLightbulb} />
          </Fab>
          <Fab
            id="crt_pl_button"
            onClick={() => onCreatePlaylist && onCreatePlaylist({ playlistName, isPrivate })}
            disabled={!isNameValid && Boolean(spotifyUser.id)}
            title={disable_create ? 'Sorry you need to be logged to spotify to be able to add playlist' : ''}
          >
            <FontAwesomeIcon icon={faSave} />
          </Fab>
          <FormControlLabel
            control={(
              <Checkbox
                id="play_list_is_private"
                value="private"
                color="primary"
                checked={isPrivate}
                onChange={({ target }) => {
                  this.setState({ isPrivate: target.checked });
                }}
              />
)}
            label="private"
          />
        </div>
      </div>
    );
  }
}

PlaylistForm.propTypes = {
  hasElements: PropTypes.bool,
  isUserLogged: PropTypes.bool,
  onStartClick: PropTypes.func,
  onCreatePlaylistClick: PropTypes.func,
  spotifyUser: PropTypes.shape(),
};

const mapStateToProps = ({ spotifyUser }) => ({
  spotifyUser,
});

export default connect(mapStateToProps)(PlaylistForm);
