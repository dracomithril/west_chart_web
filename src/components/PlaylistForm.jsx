/**
 * Created by Gryzli on 10.04.2017.
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Button, ButtonGroup, Form, FormControl, FormGroup, InputGroup } from 'react-bootstrap';
import { searchForMusic, createPlaylistAndAddTracks } from '../utils/spotify_utils';
import { getArtist_Title, weekInfo } from '../utils/utils';

import action_types from './../reducers/action_types';

export default class PlaylistForm extends Component {
  /* istanbul ignore next */
  componentDidMount() {
    console.info('component PlaylistForm did mount');
  }

  /* istanbul ignore next */
  componentWillUnmount() {
    console.info('component PlaylistForm unmounted');
  }

  onStartClick = () => {
    const { store } = this.context;
    const { selected } = this.props;
    const search = selected.map((elem, search_id) => {
      const entry = getArtist_Title(elem.link.title);
      const search_track = {
        artist: entry.artist,
        title: entry.title,
        full_title: elem.link.title,
        id: elem.id,
        search_id,
        items: [],
        selected: {},
      };
      searchForMusic(search_track, store).then(res =>
        store.dispatch({
          type: action_types.UPDATE_SINGLE_SEARCH,
          field: 'items',
          value: res.value,
          id: res.id,
        }),
      );
      return search_track;
    });
    store.dispatch({ type: action_types.UPDATE_SEARCH, search });
  };

  onGenPlaylistName = () => {
    const { store } = this.context;
    const { monday, friday } = weekInfo(new Date());

    const monday_str = monday
      .toLocaleString('en-US', { month: 'short', day: 'numeric' })
      .toUpperCase();
    const friday_str = friday
      .toLocaleString('en-US', { month: 'short', day: 'numeric' })
      .toUpperCase();

    const playlist_name = `Chart_${monday_str}-${friday_str}`;
    const list = playlist_name.split(' ').join('_');
    store.dispatch({ type: action_types.UPDATE_PLAYLIST_NAME, value: list });
  };

  onCreatePlaylist = () => {
    const { store } = this.context;
    const { search_list, sp_user, sp_playlist_name, isPlaylistPrivate } = store.getState();
    const selected = search_list
      .map(elem => (elem.selected !== undefined ? elem.selected.uri : undefined))
      .filter(elem => elem !== undefined);
    createPlaylistAndAddTracks(sp_user, sp_playlist_name, isPlaylistPrivate, selected).then(info =>
      store.dispatch({
        type: action_types.UPDATE_PLAYLIST_INFO,
        value: info,
      }),
    );
  };

  render() {
    const { store } = this.context;
    const { sp_playlist_name, isPlaylistPrivate } = store.getState();
    const { selected } = this.props;
    const disable_create = !(sp_playlist_name.length > 5 && selected.length > 0);
    const validatePlaylistName = strLength => {
      if (strLength > 8) {
        return 'success';
      }
      return strLength > 5 ? 'warning' : 'error';
    };
    return (
      <Form inline>
        <Button onClick={this.onStartClick} id="start_sp_button" bsStyle="success">
          Start
        </Button>
        <FormGroup
          style={{ margin: '1px 5px 5px 5px' }}
          controlId="play_list_name"
          validationState={validatePlaylistName(sp_playlist_name.length)}
        >
          <InputGroup style={{ maxWidth: 250 }}>
            <FormControl
              type="text"
              placeholder="playlist name"
              value={sp_playlist_name}
              onChange={e => {
                store.dispatch({
                  type: action_types.UPDATE_PLAYLIST_NAME,
                  value: e.target.value,
                });
              }}
            />
            <FormControl.Feedback />
            {/* <InputGroup.Addon><Glyphicon glyph="music"/></InputGroup.Addon> */}
          </InputGroup>
        </FormGroup>
        <ButtonGroup>
          <Button onClick={this.onGenPlaylistName} id="genName_sp_button" bsStyle="primary">
            gen. name
          </Button>
          <Button
            id="crt_pl_button"
            onClick={this.onCreatePlaylist}
            disabled={disable_create}
            bsStyle="danger"
          >
            <i className="fas fa-save" /> Playlist
          </Button>
        </ButtonGroup>
        <label htmlFor="play_list_is_private">
          <input
            type="checkbox"
            id="play_list_is_private"
            onChange={e => {
              store.dispatch({
                type: action_types.TOGGLE_IS_PRIVATE,
                value: e.target.checked,
              });
            }}
            value="private"
            checked={isPlaylistPrivate}
          />
          {'private ?'}
        </label>
      </Form>
    );
  }
}
PlaylistForm.contextTypes = {
  store: PropTypes.object,
};
PlaylistForm.propTypes = {
  selected: PropTypes.array,
};
