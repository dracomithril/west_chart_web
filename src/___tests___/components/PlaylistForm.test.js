/**
 * Created by XKTR67 on 4/19/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
import PlaylistForm from './../../components/PlaylistForm';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';

jest.mock('./../../utils/spotify_utils');
Enzyme.configure({ adapter: new Adapter() });
const initial_state = require('../data/initial_state.json');
const data = require('../data/response.json').chart;
const sinon = require('sinon');
const mockStore = configureMockStore([]);

describe('<PlaylistForm/>', () => {
  it('renders without crashing', () => {
    const store = mockStore(initial_state);
    const wrapper = shallow(
      <PlaylistForm/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
  it('start_click', (done) => {
    const sp_utils = require('../../utils/spotify_utils');
    sp_utils.searchForMusic.mockReturnValue(Promise.resolve({ id: 'zzz', value: {} }));
    const state = Object.assign({}, initial_state, { sp_playlist_name: "test_list_zzzz" });
    const store = mockStore(state);
    const wrapper = shallow(
      <PlaylistForm selected={data}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    let start_b = wrapper.find('#start_sp_button');
    start_b.simulate('click');
    expect(store.getActions().length).toBe(1);
    done();
  });
  it('createPlaylistAndAddTracks', () => {
    const sp_utils = require('../../utils/spotify_utils');
    sp_utils.createPlaylistAndAddTracks.mockReturnValue(Promise.resolve({ mock: 'mock' }));
    const state = Object.assign({}, initial_state, { sp_playlist_name: "test_list_zzzz" });
    const store = mockStore(state);
    const wrapper = shallow(
      <PlaylistForm selected={data}/>, {
        context: { store },
        childContextTypes: { store: PropTypes.object }
      }
    );

    let elem = wrapper.find('#play_list_is_private');
    let getElemByIdStub = sinon.stub();
    getElemByIdStub.withArgs('play_list_is_private').returns(elem);
    //document.getElementById = getElemByIdStub;
    expect(shallowToJson(wrapper)).toMatchSnapshot();
    let start_b = wrapper.find('#crt_pl_button');
    start_b.simulate('click');
    expect(store.getActions().length).toBe(0);
    expect(sp_utils.createPlaylistAndAddTracks.mock.calls.length).toBe(1);
  })
});
