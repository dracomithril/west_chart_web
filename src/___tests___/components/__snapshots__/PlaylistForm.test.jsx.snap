// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistForm/> createPlaylistAndAddTracks 1`] = `
<div
  className="playlist-form"
>
  <WithStyles(Button)
    color="primary"
    id="start_sp_button"
    variant="contained"
  >
    start
  </WithStyles(Button)>
  <TextField
    error={true}
    onChange={[Function]}
    placeholder="playlist name"
    required={false}
    select={false}
    style={
      Object {
        "padding": "5px",
      }
    }
    value=""
    variant="standard"
  />
  <div
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    <WithStyles(Fab)
      color="secondary"
      id="genName_sp_button"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f0eb",
              "M96.06 454.35c.01 6.29 1.87 12.45 5.36 17.69l17.09 25.69a31.99 31.99 0 0 0 26.64 14.28h61.71a31.99 31.99 0 0 0 26.64-14.28l17.09-25.69a31.989 31.989 0 0 0 5.36-17.69l.04-38.35H96.01l.05 38.35zM0 176c0 44.37 16.45 84.85 43.56 115.78 16.52 18.85 42.36 58.23 52.21 91.45.04.26.07.52.11.78h160.24c.04-.26.07-.51.11-.78 9.85-33.22 35.69-72.6 52.21-91.45C335.55 260.85 352 220.37 352 176 352 78.61 272.91-.3 175.45 0 73.44.31 0 82.97 0 176zm176-80c-44.11 0-80 35.89-80 80 0 8.84-7.16 16-16 16s-16-7.16-16-16c0-61.76 50.24-112 112-112 8.84 0 16 7.16 16 16s-7.16 16-16 16z",
            ],
            "iconName": "lightbulb",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </WithStyles(Fab)>
    <WithStyles(Fab)
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
      title="Sorry you need to be logged to spotify to be able to add playlist"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f0c7",
              "M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z",
            ],
            "iconName": "save",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </WithStyles(Fab)>
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Checkbox)
          checked={false}
          color="primary"
          id="play_list_is_private"
          onChange={[Function]}
          value="private"
        />
      }
      label="private"
    />
  </div>
</div>
`;

exports[`<PlaylistForm/> renders without crashing 1`] = `
<div
  className="playlist-form"
>
  <WithStyles(Button)
    color="primary"
    id="start_sp_button"
    variant="contained"
  >
    start
  </WithStyles(Button)>
  <TextField
    error={true}
    onChange={[Function]}
    placeholder="playlist name"
    required={false}
    select={false}
    style={
      Object {
        "padding": "5px",
      }
    }
    value=""
    variant="standard"
  />
  <div
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    <WithStyles(Fab)
      color="secondary"
      id="genName_sp_button"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f0eb",
              "M96.06 454.35c.01 6.29 1.87 12.45 5.36 17.69l17.09 25.69a31.99 31.99 0 0 0 26.64 14.28h61.71a31.99 31.99 0 0 0 26.64-14.28l17.09-25.69a31.989 31.989 0 0 0 5.36-17.69l.04-38.35H96.01l.05 38.35zM0 176c0 44.37 16.45 84.85 43.56 115.78 16.52 18.85 42.36 58.23 52.21 91.45.04.26.07.52.11.78h160.24c.04-.26.07-.51.11-.78 9.85-33.22 35.69-72.6 52.21-91.45C335.55 260.85 352 220.37 352 176 352 78.61 272.91-.3 175.45 0 73.44.31 0 82.97 0 176zm176-80c-44.11 0-80 35.89-80 80 0 8.84-7.16 16-16 16s-16-7.16-16-16c0-61.76 50.24-112 112-112 8.84 0 16 7.16 16 16s-7.16 16-16 16z",
            ],
            "iconName": "lightbulb",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </WithStyles(Fab)>
    <WithStyles(Fab)
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
      title="Sorry you need to be logged to spotify to be able to add playlist"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f0c7",
              "M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z",
            ],
            "iconName": "save",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </WithStyles(Fab)>
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Checkbox)
          checked={false}
          color="primary"
          id="play_list_is_private"
          onChange={[Function]}
          value="private"
        />
      }
      label="private"
    />
  </div>
</div>
`;

exports[`<PlaylistForm/> start_click 1`] = `
<div
  className="playlist-form"
>
  <WithStyles(Button)
    color="primary"
    id="start_sp_button"
    onClick={[Function]}
    variant="contained"
  >
    start
  </WithStyles(Button)>
  <TextField
    error={true}
    onChange={[Function]}
    placeholder="playlist name"
    required={false}
    select={false}
    style={
      Object {
        "padding": "5px",
      }
    }
    value=""
    variant="standard"
  />
  <div
    style={
      Object {
        "alignSelf": "center",
      }
    }
  >
    <WithStyles(Fab)
      color="secondary"
      id="genName_sp_button"
      onClick={[Function]}
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              352,
              512,
              Array [],
              "f0eb",
              "M96.06 454.35c.01 6.29 1.87 12.45 5.36 17.69l17.09 25.69a31.99 31.99 0 0 0 26.64 14.28h61.71a31.99 31.99 0 0 0 26.64-14.28l17.09-25.69a31.989 31.989 0 0 0 5.36-17.69l.04-38.35H96.01l.05 38.35zM0 176c0 44.37 16.45 84.85 43.56 115.78 16.52 18.85 42.36 58.23 52.21 91.45.04.26.07.52.11.78h160.24c.04-.26.07-.51.11-.78 9.85-33.22 35.69-72.6 52.21-91.45C335.55 260.85 352 220.37 352 176 352 78.61 272.91-.3 175.45 0 73.44.31 0 82.97 0 176zm176-80c-44.11 0-80 35.89-80 80 0 8.84-7.16 16-16 16s-16-7.16-16-16c0-61.76 50.24-112 112-112 8.84 0 16 7.16 16 16s-7.16 16-16 16z",
            ],
            "iconName": "lightbulb",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </WithStyles(Fab)>
    <WithStyles(Fab)
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
      title="Sorry you need to be logged to spotify to be able to add playlist"
    >
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              448,
              512,
              Array [],
              "f0c7",
              "M433.941 129.941l-83.882-83.882A48 48 0 0 0 316.118 32H48C21.49 32 0 53.49 0 80v352c0 26.51 21.49 48 48 48h352c26.51 0 48-21.49 48-48V163.882a48 48 0 0 0-14.059-33.941zM224 416c-35.346 0-64-28.654-64-64 0-35.346 28.654-64 64-64s64 28.654 64 64c0 35.346-28.654 64-64 64zm96-304.52V212c0 6.627-5.373 12-12 12H76c-6.627 0-12-5.373-12-12V108c0-6.627 5.373-12 12-12h228.52c3.183 0 6.235 1.264 8.485 3.515l3.48 3.48A11.996 11.996 0 0 1 320 111.48z",
            ],
            "iconName": "save",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        symbol={false}
        title=""
        transform={null}
      />
    </WithStyles(Fab)>
    <WithStyles(WithFormControlContext(FormControlLabel))
      control={
        <WithStyles(Checkbox)
          checked={false}
          color="primary"
          id="play_list_is_private"
          onChange={[Function]}
          value="private"
        />
      }
      label="private"
    />
  </div>
</div>
`;
